container/new_pull.bzl

```diff
@@ -0,0 +1,208 @@
+
+    # Add an empty top-level BUILD file.
+    repository_ctx.file("BUILD", "")
+
+    # TODO(xwinxu): change container_import rule to have an oci attribute to account for new OCI Image Layout
+    # currently exclude:     exports_files(["image.digest", "digest"]) in BUILD, theoretical BUILD commented out below:
+    repository_ctx.file("image/BUILD", """
+package(default_visibility = ["//visibility:public"])
+
+load("@io_bazel_rules_docker//container:import.bzl", "container_import")
+
+container_import(
+    name = "image",
+    config = "image/index.json",
+    layers = glob(["image/blobs/sha256/*"]),
+)
+    """)
+
+    args = [
+        repository_ctx.path(repository_ctx.attr._puller),
+        "-directory",
+        repository_ctx.path("image"),
+        "-os",
+        repository_ctx.attr.os,
+        "-os-version",
+        repository_ctx.attr.os_version,
+        "-os-features",
+        " ".join(repository_ctx.attr.os_features),
+        "-architecture",
+        repository_ctx.attr.architecture,
+        "-variant",
+        repository_ctx.attr.cpu_variant,
+        "-features",
+        " ".join(repository_ctx.attr.platform_features),
+    ]
+
+    # Use the custom docker client config directory if specified.
+    if repository_ctx.attr.docker_client_config != "":
+        args += ["-client-config-dir", "{}".format(repository_ctx.attr.docker_client_config)]
+
+    # TODO(xwinxu): There is currently no support for caches in rules_docker. Stay tuned...
+    # cache_dir = repository_ctx.os.environ.get("DOCKER_REPO_CACHE")
+    # if cache_dir:
+    #     if cache_dir.startswith("~/") and "HOME" in repository_ctx.os.environ:
+    #         cache_dir = cache_dir.replace("~", repository_ctx.os.environ["HOME"], 1)
+
+    #     args += [
+    #         "-cache",
+    #         cache_dir,
+    #     ]
+
+    # If a digest is specified, then pull by digest.  Otherwise, pull by tag.
+    if repository_ctx.attr.digest:
+        args += [
+            "-name",
+            "{registry}/{repository}@{digest}".format(
+                registry = repository_ctx.attr.registry,
+                repository = repository_ctx.attr.repository,
+                digest = repository_ctx.attr.digest,
+            ),
+        ]
+    else:
+        tag_ = repository_ctx.attr.tag
+        if not repository_ctx.attr.tag:
+            tag_ = "latest"
+        args += [
+            "-name",
+            "{registry}/{repository}:{tag}".format(
+                registry = repository_ctx.attr.registry,
+                repository = repository_ctx.attr.repository,
+                tag = tag_,
+            ),
+        ]
+
+    kwargs = {}
+    if "PULLER_TIMEOUT" in repository_ctx.os.environ:
+        kwargs["timeout"] = int(repository_ctx.os.environ.get("PULLER_TIMEOUT"))
+
+    result = repository_ctx.execute(args, **kwargs)
+    if result.return_code:
+        fail("Pull command failed: %s (%s)" % (result.stderr, " ".join(args)))
```

Comment:
fail("Pull command failed: %s (%s)" % (result.stderr, " ".join([str(a) for a in args])))

I think Bazel 0.27.0 would complain you are joining path with string.